crates/nargo/src/ops/foreign_calls.rs

```diff
@@ -0,0 +1,139 @@
+                Self::execute_println(&foreign_call.inputs)?;
+                Ok(foreign_call.inputs[0][0].into())
+            }
+            Some(ForeignCall::PrintlnFormat) => {
+                Self::execute_println_format(&foreign_call.inputs)?;
+                Ok(foreign_call.inputs[0][0].into())
+            }
+            Some(ForeignCall::Sequence) => {
+                let sequence_length: u128 = foreign_call.inputs[0][0].to_field().to_u128();
+
+                Ok(vecmap(0..sequence_length, Value::from).into())
+            }
+            Some(ForeignCall::ReverseSequence) => {
+                let sequence_length: u128 = foreign_call.inputs[0][0].to_field().to_u128();
+
+                Ok(vecmap((0..sequence_length).rev(), Value::from).into())
+            }
+            None => panic!("unexpected foreign call {:?}", foreign_call_name),
+        }
+    }
+
+    fn execute_println(foreign_call_inputs: &[Vec<Value>]) -> Result<(), ForeignCallError> {
+        // Fetch the abi type from the foreign call input
+        // The remaining input values should hold what is to be printed
+        let (abi_type, input_values) = fetch_abi_type(foreign_call_inputs)?;
+
+        // We must use a flat map here as each value in a struct will be in a separate input value
+        let mut input_values_as_fields =
+            input_values.iter().flat_map(|values| vecmap(values, |value| value.to_field()));
+        let decoded_value = decode_value(&mut input_values_as_fields, &abi_type)?;
+
+        let json_value = JsonTypes::try_from_input_value(&decoded_value, &abi_type)?;
+        let output_string = serde_json::to_string_pretty(&json_value)
+            .map_err(|err| ForeignCallError::InputParserError(err.into()))?;
+
+        println!("{output_string}");
+        Ok(())
+    }
+
+    fn execute_println_format(foreign_call_inputs: &[Vec<Value>]) -> Result<(), ForeignCallError> {
+        // Fetch the message from the first input
+        let (message_as_values, input_and_abi_values) =
+            foreign_call_inputs.split_first().ok_or(ForeignCallError::MissingForeignCallInputs)?;
+        // Fetch the abi type from the foreign call input
+        // The remaining input values should hold what is to be printed
+        let (abi_type, input_values) = fetch_abi_type(input_and_abi_values)?;
+
+        // Fetch the abi field count from the type to account for nested structs
+        let type_size = abi_type.field_count();
+        let input_values_chunks = input_values[0].chunks(type_size as usize);
+
+        let mut output_strings = Vec::new();
+        for input_values in input_values_chunks {
+            let input_values_as_fields = vecmap(input_values, |value| value.to_field());
+            let decoded_value = decode_value(&mut input_values_as_fields.into_iter(), &abi_type)?;
+            let json_value = JsonTypes::try_from_input_value(&decoded_value, &abi_type)?;
+            let output_string = serde_json::to_string(&json_value)
```

Comment:
I was debating between `to_string_pretty` and `to_string` here.
This is what `println_format` looks like with just `to_string`:
```
"*** println_format ***"
x: "0x05", y: "0x0a"
s1: {"x":"0x0a","y":"0x05"}, s2: {"x":"0x14","y":"0x1e"}
foo1: {"foo":"0x0f","my_struct":{"x":"0x0a","y":"0x05"}}, foo2: {"foo":"0x14","my_struct":{"x":"0x14","y":"0x1e"}}
```
While this is with `to_string_pretty`:
```
"*** println_format ***"
x: "0x05", y: "0x0a"
s1: {
  "x": "0x0a",
  "y": "0x05"
}, s2: {
  "x": "0x14",
  "y": "0x1e"
}
foo1: {
  "foo": "0x0f",
  "my_struct": {
    "x": "0x0a",
    "y": "0x05"
  }
}, foo2: {
  "foo": "0x14",
  "my_struct": {
    "x": "0x14",
    "y": "0x1e"
  }
}
```
The structs themselves are easier to read with the pretty json but the format string can start to look different than specified in the program. I'm curious what people think about this choice

Reply:
I tend to prefer the more terse version since it works better on small inputs and for larger inputs to_string_pretty can conversely clog the output too much and make it more difficult to read. A better option though would be to provide both. Like how rust offers both `"{foo:?}"` and `"{foo:#?}"`

Reply:
Ah true, I will look at proving both options. I will do the same for `println_new` as that is currently defaulted to `to_string_pretty` for individual outputs